resource-aggregate/coap-gw: implement twin synchronization states

jkralik opened the pull request:
Docs: https://github.com/plgd-dev/doc/pull/78

jkralik committed:
resource-aggregate/coap-gw: implement resource shadow synchronization status

jkralik committed:
Merge branch 'main' into jkralik/feature/reportsObservationsAreReady

jkralik committed:
introduce TwinSynchronizationStatus (#800)

* introduce TwinSynchronizationStatus

Co-authored-by: Ondrej Tomcik <ondrej.tomcik@plgd.dev>

jkralik committed:
fix tests

jkralik committed:
get CPU info from workflows

jkralik committed:
try to clean

jkralik committed:
improve test

jkralik committed:
coap-gateway: enable observation per resource

Danielius1922 commented on resource-directory/service/getEvents_test.go:
```diff
@@ -340,7 +346,7 @@ func testDeleteResourceEvents(t *testing.T, ctx context.Context, c pb.GrpcGatewa
 func TestRequestHandlerGetEventsOnCollection(t *testing.T) {
 	deviceID := test.MustFindDeviceByName(test.TestDeviceName)
-	ctx, cancel := context.WithTimeout(context.Background(), testCfg.TEST_TIMEOUT)
+	ctx, cancel := context.WithTimeout(context.Background(), testCfg.TEST_TIMEOUT*30)
```
testCfg.TEST_TIMEOUT (=30 seconds) times 30 = 15min, seems too much

jkralik replied:
fixed

jkralik committed:
fix for CR

Danielius1922 commented on resource-aggregate/pb/commands.proto:
```diff
@@ -110,6 +110,15 @@ message NotifyResourceChangedResponse {
+}
+
+message BatchNotifyResourceChangedResponse {
+    AuditContext audit_context = 3;
```
Why 3 and not 1?

jkralik replied:
fixed

jkralik committed:
coap-gateway/resource-aggrate: implement twin synchronization states

The twin synchronization state informs user about the reconciliation
after the reconnect or twin creation after the first sign-in to the plgd hub.

More info: https://docs.plgd.dev/docs/features/device-twin

Danielius1922 commented on coap-gateway/service/observation/deviceObserver.go:
```diff
@@ -183,32 +186,39 @@ func NewDeviceObserver(ctx context.Context, deviceID string, coapConn ClientConn
 		return nil, createError(err)
 	}
 
+	var notSyncedResources sync.Map
```
`notSyncedResources` is not used after line 191, that doesn't seem right

jkralik replied:
removed

Danielius1922 commented on coap-gateway/service/observation/resourcesObserver.go:
```diff
@@ -83,55 +87,95 @@ func newResourcesObserver(deviceID string, coapConn ClientConn, callbacks Resour
+		err = o.performObservationLocked([]*observedResource{obs})
+		if err != nil {
+			o.resources, _ = o.resources.removeByHref(obs.Href())
+			defer o.notifyAboutFinishTwinSynchronization(ctx)
```
in case of success (err == nil) is there a different code path which emits this notification?

jkralik replied:
Yes. In this case it will be emit by https://github.com/plgd-dev/hub/blob/jkralik/feature/reportsObservationsAreReady/coap-gateway/service/session.go#L466

Danielius1922 replied:
What if notifyContentChanged returns early with an error, won't the notification be missing then?

jkralik replied:
In this case the notification `in sync` is not set for the device and device is disconnected from coap-gw (fixed in new commit).

Danielius1922 commented on resource-aggregate/pb/commands.proto:
```diff
@@ -463,18 +473,29 @@ message ConnectionStatus {
+        SYNCING = 2; // As soon as device connects, successfully signs in and batch observe is called on device from the cloud.
+        IN_SYNC = 3; // As soon as current device resources values are received and applied to twin database. Twin was successfully reconciled after device reconnect and is kept up to date using an active subscription to device resource changes.
+    };
+    State state = 2;
```
= 1?

Danielius1922 commented on http-gateway/service/updateDeviceMetadata_test.go:
```diff
@@ -63,9 +63,10 @@ func (f *contentChangedFilter) Handle(ctx context.Context, iter eventbus.Iter) (
-			if ev.GetShadowSynchronization() == commands.ShadowSynchronization_UNSET {
-				continue
-			}
+			// TODO ????
```
remove comment or finish code if something is missing

jkralik replied:
removed

jkralik committed:
Merge branch 'main' into jkralik/feature/reportsObservationsAreReady
